syntax = "proto3";

package post;

service postService
{
  rpc postCreate (postCreateRequest) returns (postCreateResponse);
  rpc postGet (postGetRequest) returns (postGetResponse);
  rpc postUpdate (postUpdateRequest) returns (postUpdateResponse);
  rpc postDelete (postDeleteRequest) returns (postDeleteResponse);
  rpc postsListing (postsListingRequest) returns (postsListingResponse);
}

message postCreateRequest
{
  string title = 1;
  string description = 2;
  int32 user_id = 3;
  bool is_private = 4;
  repeated string tags = 5;
}

message postCreateResponse
{
  int32 post_id = 1;
}

message postGetRequest
{
  int32 post_id = 1;
  int32 user_id = 2;
}

message postGetResponse
{
  int32 post_id = 1;
  string title = 2;
  string description = 3;
  int32 user_id = 4;
  string create_time = 5;
  string last_update = 6;
  bool is_private = 7;
  repeated string tags = 8;
}

message postUpdateRequest
{
    int32 post_id = 1;
    string title = 2;
    string description = 3;
    int32 user_id = 4;
    bool is_private = 5;
    repeated string tags = 6;
}

message postUpdateResponse
{
    int32 post_id = 1;
    string title = 2;
    string description = 3;
    int32 user_id = 4;
    string create_time = 5;
    string last_update = 6;
    bool is_private = 7;
    repeated string tags = 8;
}

message postDeleteRequest
{
    int32 post_id = 1;
    int32 user_id = 2;
}

message postDeleteResponse
{
    bool success = 1;
    int32 post_id = 2;
}

message postsListingRequest
{
    int32 user_id = 3;
    int32 offset = 1;
    int32 limit = 2;
}

message postsListingResponse 
{
    repeated postGetResponse posts = 1;
}
